[DOC] Add new cluster metrics monitor supported APIs (#4525)
opensearch-project/documentation-website: _observing-your-data/alerting/monitors.md

The Alerting plugin provides four monitor types:

1. **per query**: Runs a query and generates alert notifications based on the matching criteria.
1. **per bucket**: Runs a query that evaluates trigger criteria based on aggregated values in the dataset.
1. **per cluster metrics**: Runs API requests on the cluster to monitor its health.
1. **per document**: Runs a query (or multiple queries combined by a tag) that returns individual documents that match the alert notification trigger condition.

![Monitor types in OpenSearch]({{site.url}}{{site.baseurl}}/images/monitors.png)

## Creating monitors

To create a monitor:

1. In the **OpenSearch Plugins** main menu, choose **Alerting**.
1. Choose **Create monitor**.
1. Enter the **Monitor details**, including monitor type, method, and schedule.  
1. Select a data source from the dropdown list.
1. Define the metrics in the Query section.

> 1. Add a [trigger](insert-link).

Reviewer: Insert link :)

1. Select **Create**.

> The maximum number of monitors you can create is 1,000. You can change the default maximum number of alerts for your cluster by calling the [cluster settings API]({{site.url}}{{site.baseurl}}/observing-your-data/alerting/settings/) `plugins.alerting.monitor.max_monitors`.

Reviewer: ```suggestion
The maximum number of monitors you can create is 1,000. You can change the default maximum number of alerts for your cluster by updating the `plugins.alerting.monitor.max_monitors` setting using the [cluster settings API]({{site.url}}{{site.baseurl}}/observing-your-data/alerting/settings/).
```

#### Create sender type

You need to specify an email account from which the Alerting plugin can send notifications.

To configure a sender email:

1. Choose **Email** as the channel type.
2. Choose **Sender type**, and enter sender name, and default recipients if creating a recipient group.
1. Enter the email address, SMTP host (for example, `smtp.gmail.com` for a Gmail account), and the port.
1. Choose an encryption method, or use the default value of **None**. However, most email providers require SSL or TLS, which require a username and password in OpenSearch keystore. Refer to [Authenticate sender account](#authenticate-sender-account) to learn more.
1. Choose **Save** to save the configuration and create the sender. You can create a sender even before you add your credentials to the OpenSearch keystore. However, you must [authenticate each sender account](#authenticate-sender-account) before you use the channel to send your alert.

> You can reuse senders across many different destinations, but each destination only supports one sender.

Author: @AWSHurneyt Please confirm this sentence is accurate. Destinations are now notification channels, so we need to rephrase this sentence accordingly.

Reviewer 2: @vagimeli Yes, this is accurate as of v2.0. Thank you!

`ctx.monitor.inputs.search.indices` | Array | An array that contains the indexes the monitor observes.
`ctx.monitor.inputs.search.query` | N/A | The definition used to define the monitor.

## Creating per document monitors
Introduced 2.0
{: .label .label-purple }

Per document monitors allow you to define up to 10 queries that compare a selected field with a desired value. You can define supported field data types using the following operators:

- `is` 
- `is not`
- `is greater than`
- `is greater than equal`
- `is less than`
- `is less than equal`

You can query each trigger using up to 10 tags, adding the tag as a single trigger condition instead of specifying a single query. The Alerting plugin processes the trigger conditions from all queries as a logical `OR` operation, so if any of the query conditions are met, it triggers an alert. The Alerting plugin then tells the Notifications plugin to send the alert notification to a channel.

> The Alerting plugin also creates a list of document findings that contains metadata about which document matches each query. Security analytics can use the document findings data to keep track of and analyze the query data separately from the alert processes.

Reviewer: This is referring to the Security Analytics plugin, correct?
```suggestion
The Alerting plugin also creates a list of document findings that contains metadata about which document matches each query. Security Analytics can use the document findings data to keep track of and analyze the query data separately from the alert processes.
```

The Alerting API provides a _document-level monitor_ that programmatically accomplishes the same function as the _per document monitor_ in the OpenSearch Dashboards. See [Document-level monitors]({{site.url}}{{site.baseurl}}/monitoring-plugins/alerting/api/#document-level-monitors) to learn more.
{: .note}

### Searching document findings

> When a per document monitor executes a query that matches a document in an index, a finding is created. OpenSearch provides a findings index, `.opensearch-alerting-finding*`, that contains findings data for all per document monitor queries. You can search the findings index with the Alerting API search operation. See [Search for monitor findings]({{site.url}}{{site.baseurl}}/monitoring-plugins/alerting/api/#search-for-monitor-findings) for more information.

Author: Replaced with "runs"

The following metadata is provided for each document findings entry:

* **Document** – The document ID and index name. For example: `Re5akdirhj3fl | test-logs-index`.
* **Query** – The query name that matched the document.
* **Time found** – The timestamp that indicates when the document was found during the runtime.

> While you can configure an alert notification for each finding, it is not recommended unless rules are well defined to prevent a large volume of findings in a high ingestion cluster.

Reviewer: ```suggestion
To prevent a large volume of findings in a high ingestion cluster, configuring alert notifications for each finding is not recommended unless rules are well defined.
{: .important}
```

## Creating cluster metrics monitors

In addition to monitoring conditions for indexes, the Alerting plugin allows monitoring conditions for clusters. Alerts can be set by cluster metrics to watch for the following conditions:

- Cluster health status reaches yellow or red.

> - Cluster-level metrics, such as CPU usage and JVM memory usage, reach specified thresholds.

Author: Good suggestion. Revised

- Node-level metrics, such as available disk space, JVM memory usage, and CPU usage, reach a specified threshold.
- Total number of documents stores reaches a specified amount.

To create a cluster metrics monitor:

1. In the **OpenSearch Plugins** main menu, select **Alerting**. 
1. Select **Monitors**, then **Create monitor**.
1. Select **Per cluster metrics monitor**.
1. In the Query section, select **Request type** from the dropdown.
1. To filter the API response to use only certain path parameters, enter those parameters in the **Query parameters** field. Most APIs that can be used to monitor cluster status support path parameters, as described in their respective documentation (for example, comma-separated lists of index names).

> 1. In the Triggers section, define the conditions to trigger an alert. The trigger condition autopopulates a Painless `ctx` variable. For example, a cluster monitor watching for cluster stats uses the trigger condition `ctx.results[0].indices.count <= 0`, which triggers an alert based on the number of indexes returned by the query. For more specificity, add any additional Painless conditions supported by the API. To preview the condition response, select **Preview condition response**.

Author: Accepted

> 1. In the Triggers section, define the conditions to trigger an alert. The trigger condition autopopulates a Painless `ctx` variable. For example, a cluster monitor watching for cluster stats uses the trigger condition `ctx.results[0].indices.count <= 0`, which triggers an alert based on the number of indexes returned by the query. For more specificity, add any additional Painless conditions supported by the API. Tp preview the condition response, select **Preview condition response**.

Reviewer: ```suggestion
1. In the Triggers section, define the conditions to trigger an alert. The trigger condition autopopulates a Painless `ctx` variable. For example, a cluster monitor watching for cluster stats uses the trigger condition `ctx.results[0].indices.count <= 0`, which triggers an alert based on the number of indexes returned by the query. For more specificity, provide additional Painless conditions supported by the API. To preview the condition response, select **Preview condition response**.
```

### Painless triggers

Painless scripts define triggers for cluster metrics monitors, similar to query or bucket-level monitors that are defined using the extraction query definition option. Painless scripts comprise at least one statement and any additional functions you want to run. The cluster metrics monitor supports up to **ten** triggers.

> In the following example, a JSON object creates a trigger that sends an alert when the cluster health is yellow. `script` points the `source` to the Painless script `ctx.results[0].status == \"yellow\`.

Reviewer: ```suggestion
In the following example, a JSON object defines a trigger that sends an alert when the cluster health is yellow. `script` points the `source` to the Painless script `ctx.results[0].status == \"yellow\`.
```

}
```

### Limitations

The cluster metrics monitor has the following limitations:

- Monitors cannot be created for remote clusters.
- The OpenSearch cluster must be in a state where an index's conditions can be monitored and actions can be run against the index.
- Removing resource permissions from a user does not prevent that user’s preexisting monitors for that resource from running.

> - Users with permissions to create monitors are not blocked from creating monitors for resources for which they do not have permissions; however, those monitors do not run.

Author: Revised.

>     **Note**: Anomaly detection is available only if you are defining a per query monitor.

Author: @AWSHurneyt Is this still an accurate statement?

Reviewer 2: @vagimeli Yes, this is still accurate.
